### PR TITLE
SONARJAVA-1993 Handle eclipse NonNull and Nullable annotations

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/resolve/BytecodeAnnotationVisitor.java [hunk 1]
```diff
@@ -53,24 +52,28 @@ public AnnotationVisitor visitAnnotation(String name, String desc) {
...
+    getSymbol(desc).members().lookup(value).stream()
+      .filter(symbol -> symbol.isKind(JavaSymbol.VAR))
+      .forEach(symbol -> addValue(name, symbol));
   }
 
   @Override
   public AnnotationVisitor visitArray(final String name) {
     final List<Object> valuesList = Lists.newArrayList();
-    //TODO handle arrays of annotation and arrays of enum values.
+    // TODO handle arrays of annotation
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/BytecodeAnnotationVisitor.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&branch=SONARJAVA-1993&issues=AWEEw-oCf6JiG_HArtjv&open=AWEEw-oCf6JiG_HArtjv)

[ISSUE_KEY]: AWEEw-oCf6JiG_HArtjv

File: java-frontend/src/test/java/org/sonar/java/se/NullableAnnotationUtilsTest.java [hunk 2]
```diff
@@ -21,26 +21,30 @@
...
 
 public class NullableAnnotationUtilsTest {
 
-  private static SemanticModel semanticModel;
+  private SemanticModel semanticModel;
 
-  @BeforeClass
-  public static void setUp() {
+  @Before
+  public void setUp() {
```
Comment (benzonico) [inline java-frontend/src/test/java/org/sonar/java/se/NullableAnnotationUtilsTest.java]: This change implies parsing and semantic computation for every test of this class. This seems quite some overhead.

File: java-frontend/src/main/java/org/sonar/java/se/NullableAnnotationUtils.java [hunk 3]
```diff
@@ -61,10 +64,39 @@ private static boolean isAnnotatedWithOneOf(Symbol symbol, Set<String> annotatio
...
+      return valuesForAnnotation(enclosingClassSymbol, annotation);
+    }
+    JavaSymbol.PackageJavaSymbol packageSymbol = methodSymbol.packge();
+    if (isAnnotatedWith(packageSymbol, annotation)) {
+      return valuesForAnnotation(packageSymbol, annotation);
+    }
+    return Collections.emptyList();
+  }
+
+  private static List<SymbolMetadata.AnnotationValue> valuesForAnnotation(Symbol symbol, String annotation) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/NullableAnnotationUtils.java]: I think we should inline that one. It does not bring any clarity to the code.

File: java-frontend/src/main/java/org/sonar/java/se/NullableAnnotationUtils.java [hunk 4]
```diff
@@ -92,11 +119,34 @@ public static boolean isGloballyAnnotatedWith(MethodTree methodTree, String anno
...
+      case IDENTIFIER:
+        symbol = ((IdentifierTree) defaultLocation).symbol();
+        break;
+      case MEMBER_SELECT:
+        symbol = ((MemberSelectExpressionTree) defaultLocation).identifier().symbol();
+        break;
+      case NEW_ARRAY:
+        return ((NewArrayTree) defaultLocation).initializers().stream().anyMatch(expr -> containsDefaultLocation(expr, target));
+      default:
+        // other tree which may be used in annotations ?
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/NullableAnnotationUtils.java]: this comment is rather odd... I guess the message in the exception is sufficient.

File: java-frontend/src/main/java/org/sonar/java/resolve/BytecodeAnnotationVisitor.java (hunk 1 above)
Comment (m-g-sonar) [reply java-frontend/src/main/java/org/sonar/java/resolve/BytecodeAnnotationVisitor.java]: I won't fix this remaining part, as I just improved the current handling to cover enums. Marked as "Confirmed".

**Pull request merged**